Add FilterLineEdit to unify editor filter field navigation

KoBeWi opened the pull request:
Follow-up to #96400
The repetitive code for handling input forwarding was extracted to a new class.
Also tweaked some includes.

KoBeWi committed:
Add FilterLineEdit to unify editor filter field navigation

KoBeWi commented on editor/property_selector.cpp:
```diff
@@ -40,32 +41,6 @@ void PropertySelector::_text_changed(const String &p_newtext) {
-			search_options->gui_input(key);
-			search_box->accept_event();
-
-			TreeItem *root = search_options->get_root();
-			if (!root->get_first_child()) {
-				return;
-			}
-
-			TreeItem *current = search_options->get_selected();
-
-			TreeItem *item = search_options->get_next_selected(root);
-			while (item) {
-				item->deselect(0);
-				item = search_options->get_next_selected(item);
-			}
-
-			current->select(0);
```
No idea what the extra code was for. Removing it doesn't seem to have changed anything.

kitbdev replied:
Looks like it is trying to deselect all but the first selected, but its already in single select mode, so it isn't needed.

Mickeon commented:
You basically took this short [Rocketchat](https://chat.godotengine.org/channel/editor?msg=oXduvpHyW5brFasnB) snippet at heart. This new class could be an outstanding starting point.

KoBeWi commented:
Not really, this was inspired by https://github.com/godotengine/godot/pull/96400#pullrequestreview-2301351811
For searching improvements, #98278 might be more helpful.

a-johnston commented:
I think the most related code in that PR is with `QuickOpenResultContainer` where the parent container determines if a given event goes to the LineEdit or grid/list navigation https://github.com/godotengine/godot/blob/562c666/editor/gui/editor_quick_open_dialog.cpp#L465-L487

If `FilterLineEdit` had additional configuration for passing along for example left/right with a specific modifier (or lack of modifier) it could be used in `QuickOpenResultContainer` (assuming we then update the config when grid/list views are selected). I don't know if that abstraction would be worthwhile (how often do we need to delegate between list/grid view while a text field is active?) but it's at least relevant to consider.

IMO `FuzzySearch` benefits from not being tied explicitly to a node but if there is enough demand for a `LineEdit` or `ItemList` subclass which can do that sort of filtering, it maybe still makes sense to add a relatively lightweight class that then calls into `FuzzySearch`. Or maybe it's good enough to just be able to call into `FuzzySearch` off of text_changed/text_submitted like with https://github.com/a-johnston/godot/blob/3ac043c/editor/gui/editor_quick_open_dialog.cpp#L413-L423

The mentioned features in that rocketchat message could be valuable to add to FuzzySearch in any case - I like the idea of quoted tokens always using the exact match and normalizing accented characters etc.

Maran23 commented on editor/create_dialog.cpp:
```diff
@@ -785,7 +790,7 @@ CreateDialog::CreateDialog() {
+	search_box = memnew(FilterLineEdit);
 	search_box->set_clear_button_enabled(true);
 	search_box->set_h_size_flags(Control::SIZE_EXPAND_FILL);
 	search_box->connect(SceneStringName(text_changed), callable_mp(this, &CreateDialog::_text_changed));
```
I think you forgot to remove the line below, which is now in the `FilterLineEdit`.
`search_box->connect(SceneStringName(gui_input), callable_mp(this, &CreateDialog::_sbox_input));`

KoBeWi replied:
`_sbox_input()` has some custom code for collapsing items, so the signal is still needed. I just forgot to remove the part handled by FilterLineEdit; fixed now.

Maran23 commented:
Branch has conflicts. Otherwise looks good to me and a nice continuation from my cleanup.

Makes sense that the `CreateDialog` has handling for opening or closing the `TreeItem`s with `Space`, since Classes can have no whitespace in the name, `Space` is not needed and therefore can be consumed.

Maran23 reviewed (approved):
needs another rebase, LGTM otherwise

kitbdev commented on editor/script/script_editor_plugin.cpp:
```diff
@@ -448,23 +438,28 @@ void ScriptEditorQuickOpen::_bind_methods() {
 	search_box->connect(SceneStringName(text_changed), callable_mp(this, &ScriptEditorQuickOpen::_text_changed));
-	search_box->connect(SceneStringName(gui_input), callable_mp(this, &ScriptEditorQuickOpen::_sbox_input));
+
 	search_options = memnew(Tree);
```
```suggestion
	search_options = memnew(Tree);
	search_box->set_forward_control(search_options);
```

kitbdev reviewed (commented):
Makes sense.

This can also be done in EditorHelpSearch.

We can also have a default `set_clear_button_enabled` to true for all of these, it makes sense for all of these filters.
Same with the right icon, "Search" is a good default here.

I think it also makes sense to use this in FileSystemDock, SceneTreeDock, GroupsEditor, ProjectManager, ConnectionsDock, ConnectDialog (nodes and methods), SpriteFramesEditor, ThemeItemImportTree, ScriptEditor `filter_methods` (the `filter_scripts` would need more work), EditorSettingsDialog, ProjectSettingsEditor, and TreeSortAndFilterBar. These could be done here or in another PR.
Or maybe we can use TreeSortAndFilterBar instead?

KoBeWi commented:
Applied suggestions and added default icon/clear button to FilterLineEdit (and removed related code that was setting them externally).

As for adding the FilterLineEdit in more places, I'd rather do it in a follow-up. This PR was supposed to unify existing search boxes with custom input.

kitbdev commented on editor/script/script_editor_plugin.cpp:
```diff
@@ -448,23 +432,25 @@ void ScriptEditorQuickOpen::_bind_methods() {
-	vbc->add_margin_child(TTRC("Matches:"), search_options, true);
-	set_ok_button_text(TTRC("Open"));
-	get_ok_button()->set_disabled(true);
-	register_text_enter(search_box);
```
This `register_text_enter` was deleted, but it should be kept.

kitbdev reviewed (commented):
CreateDialog has a `search_box->set_right_icon` that can be removed.

The GridMapEditor should also accept left/right inputs since it is a grid. But it is fine for now since just up/down is the current behavior.

Repiteo commented:
Needs rebase
